File: c/670/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

size_t	getDigits(int n) {
	size_t	i;

	i = 1;
	while (n /= 10)
		i++;
	return (i);
}

char			*itoa(int n) {
	char		*str;
	size_t		digits;
	long int	num;

	num = n;
	digits = getDigits(n);
	if (n < 0) {
		num *= -1;
		digits++;
	}
	if (!(str = (char *)malloc(sizeof(char) * (digits + 1)))) {
		return (NULL);
    }
	*(str + digits) = 0;
	while (digits--) {
		*(str + digits) = num % 10 + '0';
		num = num / 10;
	}
	if (n < 0) {
		*(str + 0) = '-';
    }
	return (str);
}

int maximumSwap(int num) {
    char *buf = itoa(num);

    int last[10];
    for (int i = 0; i < 10; i++) {
        last[i] = -1;
    }

    for (int i = 0; buf[i]; i++) {
        last[buf[i] - '0'] = i;
    }

    for (int i = 0; buf[i]; i++) {
        for (int d = 9; d > buf[i] - '0'; d--) {
            if (last[d] > i) {
                char temp = buf[i];
                buf[i] = buf[last[d]];
                buf[last[d]] = temp;
                return atoi(buf);
            }
        }
    }
    
    return num;
}

// Testing
int main() {

    int testCases[] =       {2736, 9973, 1234, 4321, 0, 9, 987654321, 1234, 1993};
    int expectedResults[] = {7236, 9973, 4231, 4321, 0, 9, 987654321, 4231, 9913};

    size_t numTests = sizeof(testCases) / sizeof(testCases[0]);
    
    for (size_t i = 0; i < numTests; i++) {
        int result = maximumSwap(testCases[i]);
        printf(">> Test case [%d]: maximumSwap(%d) = %d; Expected = %d\n", (int)i + 1, testCases[i], result, expectedResults[i]);
        printf("%s\n", result == expectedResults[i] ? "[+] Success !" : "[-] Failed !" );
    }

    return 0;
}

File: typescript/670/main.ts
function maximumSwap(num: number): number {
    let digits: number[] = num.toString().split('').map(Number);
    const strlen = digits.length;

    let index1:number = -1;
    let index2:number = -1;
    let maxDigit:number = -1;
    let maxIndex:number = -1;

     for (let i = strlen - 1; i >= 0; i--) {
        if (digits[i] > maxDigit) {
            maxDigit = digits[i];
            maxIndex = i;
        }
        else if (digits[i] < maxDigit) {
            index1 = maxIndex;
            index2 = i;
        }
    }

    if (index1 > -1) {
        const temp:number = digits[index1];
        digits[index1] = digits[index2];
        digits[index2] = temp;
        return parseInt(digits.join(''), 10);
    }

    return num;
};

// Testing
const testCases = [
    { nums: 67345,      expectedNum: 76345 },
    { nums: 31455,      expectedNum: 51453 },
    { nums: 123214,     expectedNum: 423211 },
    { nums: 12,         expectedNum: 21 },
    { nums: 21,         expectedNum: 21 },
    { nums: 1,          expectedNum: 1 },
    { nums: 191,        expectedNum: 911 }
];

testCases.forEach(({ nums, expectedNum }, index) => {
    let result: number = maximumSwap(nums);
    console.log(`Testcase [${index}] >> ${result == expectedNum ? "[+] Success !" : "[+] Failed"}`)
});